### PR TITLE
pypy{27,38,39}: refactor derivation

### PR DESCRIPTION
**Pull request by thiagokokada**
###### Description of changes

- Remove code related to i686 archs (not supported anymore since #209100)
- Split non-install steps in installPhase in installCheckPhase/preFixup
- Add missing pre/post hooks
- Remove global `with lib;` usage (https://github.com/NixOS/nixpkgs/issues/208242)

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by thiagokokada** [comment]
Result of `nixpkgs-review pr 209675` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>cvs2svn</li>
    <li>pypy</li>
    <li>pypy3</li>
    <li>pypy38</li>
  </ul>
</details>

**Comment by thiagokokada** [comment]
Result of `nixpkgs-review pr 209675` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>tests.writers</li>
  </ul>
</details>
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>cvs2svn</li>
    <li>pypy (pypy2 ,pypy27)</li>
    <li>pypy3 (pypy39)</li>
    <li>pypy38</li>
  </ul>
</details>

**Comment by thiagokokada** [comment]
CC @AndersonTorres

**Comment by thiagokokada** [comment]
@ofborg build pypy27 pypy38 pypy39

File: pkgs/development/interpreters/python/pypy/default.nix [hunk 1]
```diff
@@ -96,14 +92,43 @@ in with passthru; stdenv.mkDerivation rec {
...
+
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p $out/{bin,include,lib,${executable}-c}
+
+    cp -R {include,lib_pypy,lib-python,${executable}-c} $out/${executable}-c
+    cp lib${executable}-c${stdenv.hostPlatform.extensions.sharedLibrary} $out/lib/
+    ln -s $out/${executable}-c/${executable}-c $out/bin/${executable}
+    ${lib.optionalString isPy39OrNewer "ln -s $out/bin/${executable}-c $out/bin/pypy3"}
```
Comment (wegank) [inline pkgs/development/interpreters/python/pypy/default.nix:118]: While debugging #209699, I observed that this symlink is not created in pypy39 on aarch64-darwin. Maybe there is a typo?

```suggestion
    ${lib.optionalString isPy39OrNewer "ln -s $out/bin/${executable} $out/bin/pypy3"}
```

**Review by AndersonTorres — COMMENTED** [review]
LGTM. Waiting my local machine.

File: pkgs/development/interpreters/python/pypy/default.nix (hunk 1 above)
Comment (thiagokokada) [reply pkgs/development/interpreters/python/pypy/default.nix]: Yeah, you're right. And this is probably broken in all archs.

Comment (thiagokokada) [reply pkgs/development/interpreters/python/pypy/default.nix]: Fixed in [7e071f9](https://github.com/NixOS/nixpkgs/pull/209675/commits/7e071f9e732ad7ef2988165ff63a6f2d99d6d0b0).

**Comment by thiagokokada** [comment]
https://github.com/ofborg build pypy27 pypy38 pypy39

**Comment by thiagokokada** [comment]
@ofborg build pypy27 pypy38 pypy39

**Comment by wegank** [comment]
Oops, doesn't work on aarch64-darwin. I guess #209699 was too immature.

**Comment by AndersonTorres** [comment]
LGTM.

**Review by AndersonTorres — APPROVED** [review]

**Comment by thiagokokada** [comment]
@ofborg pypy27 pypy38 pypy39

**Comment by thiagokokada** [comment]
Result of `nixpkgs-review pr 209675` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>cvs2svn</li>
    <li>pypy</li>
    <li>pypy3</li>
    <li>pypy38</li>
  </ul>
</details>

**Comment by thiagokokada** [comment]
@ofborg build pypy27 pypy38 pypy39

**Review by wegank — APPROVED** [review]
```
$ ls result/bin 
pypy3   pypy3.9

$ result/bin/pypy3
Python 3.9.16 (feeb267ead3e6771d3f2f49b83e1894839f64fb7, Jan 08 2023, 19:32:57)
[PyPy 7.3.11 with GCC Clang 11.1.0] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>>> 

$ result/bin/pypy3.9 
Python 3.9.16 (feeb267ead3e6771d3f2f49b83e1894839f64fb7, Jan 08 2023, 19:32:57)
[PyPy 7.3.11 with GCC Clang 11.1.0] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>>> 
```

**Comment by thiagokokada** [comment]
Timeout in `aarch64-linux`, let me try again.

@ofborg build pypy27 pypy38 pypy39

**Comment by wegank** [comment]
Looks good on ofborg.
